File: tests_python/test_synthetic_pinnacle.py
import json
from pathlib import Path

import pytest

from app.synthetic_pinnacle import (
    build_team_pair_features,
    load_synthetic_pinnacle_bundle,
    predict_synthetic_pinnacle_from_features,
    predict_synthetic_pinnacle_quotes,
)


ARTIFACT = Path("artifacts/modeling-research/synthetic-pinnacle-direct-market-v2")


def test_r_python_synthetic_pinnacle_parity():
    bundle = load_synthetic_pinnacle_bundle(
        ARTIFACT / "synthetic-pinnacle-direct-bundle.json"
    )
    fixture = json.loads((ARTIFACT / "parity-fixture.json").read_text(encoding="utf-8"))
    quote = fixture["soft_quotes"][0]
    result = predict_synthetic_pinnacle_from_features(
        fixture["features"], fixture["features"]["league_model"],
        quote["line"], quote["odds_over"], quote["odds_under"], bundle,
    )
    for field in (
        "predicted_final_line", "predicted_final_probability_over",
        "predicted_final_odds_over", "predicted_final_odds_under",
    ):
        assert result[field] == pytest.approx(
            fixture["expected"][field], abs=1e-9
        )
    assert result["automatic_betting_approved"] is False


def test_pair_features_are_side_invariant():
    ratings = {}
    for window in ("season", "last15"):
        ratings.update({
            f"{window}_team_games": 10,
            f"{window}_attack_ratio": 1.1,
            f"{window}_concession_ratio": 0.9,
            f"{window}_kpm_ratio": 1.1,
            f"{window}_dpm_ratio": 0.9,
            f"{window}_duration_ratio": 1.0,
            f"{window}_total_kills_sd_ratio": 1.0,
            f"{window}_league_kills_per_map": 15,
            f"{window}_league_deaths_per_map": 15,
            f"{window}_league_duration": 30,
        })
    first = {"hist_pace": 1.0, "ratings": ratings}
    second = {"hist_pace": 1.2, "ratings": {**ratings, "last15_attack_ratio": 0.8}}
    assert build_team_pair_features(first, second, 2) == build_team_pair_features(
        second, first, 2
    )


def test_contract_has_no_prohibited_market_inputs():
    bundle = load_synthetic_pinnacle_bundle(
        ARTIFACT / "synthetic-pinnacle-direct-bundle.json"
    )
    assert bundle["prohibited_features"] == [
        "moneyline", "side", "soft_line", "soft_odds", "draft"
    ]
    assert not any(
        prohibited in feature.lower()
        for feature in bundle["feature_names"]
        for prohibited in ("moneyline", "odds", "soft", "blue", "red", "side")
    )


def test_three_soft_quotes_use_the_same_direct_pinnacle_forecast():
    bundle = load_synthetic_pinnacle_bundle("app_data/synthetic_pinnacle_bundle.json")
    team_ids = list(bundle["team_roster_features"])
    ratings = {}
    for window in ("season", "last15"):
        ratings.update({
            f"{window}_team_games": 20,
            f"{window}_attack_ratio": 1.0,
            f"{window}_concession_ratio": 1.0,
            f"{window}_kpm_ratio": 1.0,
            f"{window}_dpm_ratio": 1.0,
            f"{window}_duration_ratio": 1.0,
            f"{window}_total_kills_sd_ratio": 1.0,
            f"{window}_league_kills_per_map": 15,
            f"{window}_league_deaths_per_map": 15,
            f"{window}_league_duration": 30,
        })
    first = {"team_id": team_ids[0], "hist_pace": 1.0, "ratings": ratings}
    second = {"team_id": team_ids[1], "hist_pace": 1.0, "ratings": ratings}
    quotes = [
        {"line": 24.5, "odds_over": 1.90, "odds_under": 1.90},
        {"line": 25.5, "odds_over": 2.00, "odds_under": 1.80},
        {"line": 26.5, "odds_over": 2.10, "odds_under": 1.72},
    ]
    results = predict_synthetic_pinnacle_quotes(
        first, second, bundle["league_levels"][0], 1, quotes, bundle
    )
    assert len(results) == 3
    assert len({result["predicted_final_line"] for result in results}) == 1
    assert results[0]["probability_over"] > results[2]["probability_over"]

File: tests/testthat/test-synthetic-pinnacle.R
synthetic_row <- function(swapped = FALSE) {
  row <- data.frame(map_number = 2, pace = 1.1)
  for (window in c("season", "last15")) {
    for (side in c("blue", "red")) {
      value <- if (side == "blue") 1.1 else 0.9
      row[[paste(side, window, "team_games", sep = "_")]] <- 12
      row[[paste(side, window, "attack_ratio", sep = "_")]] <- value
      row[[paste(side, window, "concession_ratio", sep = "_")]] <- 2 - value
      row[[paste(side, window, "kpm_ratio", sep = "_")]] <- value
      row[[paste(side, window, "dpm_ratio", sep = "_")]] <- 2 - value
      row[[paste(side, window, "duration_ratio", sep = "_")]] <- value
      row[[paste(side, window, "total_kills_sd_ratio", sep = "_")]] <- value
      row[[paste(side, window, "league_kills_per_map", sep = "_")]] <- 15
      row[[paste(side, window, "league_deaths_per_map", sep = "_")]] <- 15
      row[[paste(side, window, "league_duration", sep = "_")]] <- 30
    }
  }
  if (swapped) {
    blue <- grep("^blue_", names(row), value = TRUE)
    red <- sub("^blue_", "red_", blue)
    temporary <- row[blue]
    row[blue] <- row[red]
    row[red] <- temporary
  }
  row
}

test_that("synthetic Pinnacle features do not depend on side assignment", {
  first <- build_synthetic_pinnacle_features(synthetic_row())
  second <- build_synthetic_pinnacle_features(synthetic_row(TRUE))
  expect_equal(first, second)
})

test_that("synthetic Pinnacle features contain no market input", {
  features <- build_synthetic_pinnacle_features(synthetic_row())
  expect_false(any(grepl("odds|moneyline|soft|blue|red", names(features))))
  expect_true(all(is.finite(as.matrix(features))))
})

test_that("direct synthetic Pinnacle bundle predicts line and prices", {
  path <- testthat::test_path(
    "..", "..", "app_data", "synthetic_pinnacle_bundle.json"
  )
  skip_if_not(file.exists(path))
  bundle <- jsonlite::read_json(path, simplifyVector = TRUE)
  expect_identical(bundle$target_mode, "direct_line_price")
  expect_identical(bundle$selected_feature_family, "structural_without_roster")
  expect_identical(
    bundle$roster_challenger_status,
    "rejected_on_confirmation_line_mae"
  )
  expect_true(all(c("line_model", "price_model", "hold_model") %in% names(bundle)))
  expect_false(any(grepl(
    "moneyline|odds|soft|side|blue|red|draft",
    bundle$feature_names,
    ignore.case = TRUE
  )))
})
